### DISCUSSION TITLE
job parallelization cluster

### DISCUSSION THREAD

**Discussion by jasper12138** [SSCHAcode/python-sscha#99] [Ideas]
Hi developers,
Here I just write a little advice for worker on remote cluster
Typically for 1000 configurations, it needs parallelization on cluster. 
It can be realized by a control module similar in crystal structure prediction, such as AIRSS and USPEX code.
The workflow is that each population submits (by control node) the configurations by parallelization (calculation nodes) with and wait every job done to process sscha code (control node).
regards,
Cong

**Comment by jasper12138** [comment]
Hi developers,
Since I can not repeat your Tutorial on quantum relaxations on H3S on cluster, I try to imitate the auto relaxation by manually repeat PbTe tutorial with cluster and even write a control script for it.
1. I set my_hpc=None, but the path of pw.x is not specified in the tutorial. What is the relevant parameter?
2. in the frequency vs steps figure, what is the step mean? N_configs = 200, max_pop = 6, but step > 4000.
Regards,
Cong

**Reply by mesonepigreco** [reply]
Hi, 
thank you for the feedback.
We implement an automatic parallelization to work with a cluster.
In this case, the SSCHA code needs to run on a local machine able to connect via ssh to the cluster, and the DFT calculations are automatically submitted to the cluster in batches.

The official guide has a section on how the automatic submission works:
https://github.com/SSCHAcode/python-sscha/blob/master/python-sscha.pdf

The cluster submission works with quantum espresso. In principle, it is possible to work with any code without modifying the source; however, this function is not yet documented (it will probably be in the next release). 

This guide also has a tutorial on automatic relaxation using an ASE calculator for gold, which is superfast also on a laptop, which I guess could be very helpful to answer your other questions.


Regarding the my_hpc in the PbTe tutorial, this is the way in which you can setup automatically the calculation on the cluster, if you set it to None, the code tries to run locally using the standard ASE Espresso calculator, that needs anyway to be provided for the automatic calculation.
The path to the binary is stored in that case in the environmental variable ASE_ESPRESSO_COMMAND which, by default, is pw.x -i PREFIX.pwi > PREFIX.pwo. You can modify it to set a mpirun to run in parallel before running the script:

```
ASE_ESPRESSO_COMMAND="mpirun -np 4 pw.x -npool 2 -i PREFIX.pwi >  PREFIX.pwo" python myscript.py
```

Or specify the full path to the pw.x executable.

In the frequency vs step, the step is the iteration of the free energy minimization algorithm, it does not depend on the number of configurations. The population, instead, is the iteration at which a new ensemble is computed. Since the code uses a reweighing, it can perform minimization steps without recomputing the ensemble each time, so we distinguish between steps of the algorithm (steps) and iterations at which we recompute the full ensemble (populations).

I hope this can help clarify a bit! Let us know if you have any other questions.
Bests,
Lorenzo

**Reply by jasper12138** [reply]
thank you. It is clear for me now.
